library/nist_kw.c

```diff
@@ -0,0 +1,650 @@
+/**
+ * \file kw.c
+ *
+ * \brief This file provides an API key wrapping(KW)
+ *        and key wrapping with padding(KWP) as defined in NIST SP800-38F
+ *        https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-38F.pdf
+ *
+ * Key-wrapping specifies a deterministic authenticated-encryption mode of operation,
+ * according to <em>NIST SP 800-38F: Recommendation for Block Cipher Modes of Operation:
+ * Methods for Key Wrapping</em>.
+ * Its purpose is to protect cryptographic keys.
+ * It uses units of 8 Bytes called semiblocks. The minimal number of input semiblocks is:
+ * <ul><li>For KW mode: 2 semiblocks.</li>
+ * <li>For KWP mode: 1 semiblock.</li></ul>
+ *
+ */
+/*
+ *  Copyright (C) 2018, Arm Limited (or its affiliates), All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of Mbed TLS (https://tls.mbed.org)
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#if defined(MBEDTLS_NIST_KW_C)
+
+#include "mbedtls/nist_kw.h"
+#include "mbedtls/platform_util.h"
+
+#include <stdint.h>
+#include <string.h>
+
+#if defined(MBEDTLS_SELF_TEST) && defined(MBEDTLS_AES_C)
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdio.h>
+#define mbedtls_printf printf
+#endif /* MBEDTLS_PLATFORM_C */
+#endif /* MBEDTLS_SELF_TEST && MBEDTLS_AES_C */
+
+#if !defined(MBEDTLS_NIST_KW_ALT)
+
+#define KW_SEMIBLOCK_LENGTH    8
+
+/*! The 64-bit default ICV for KW mode. */
```

Comment:
This is the first time we've used the abbreviation "ICV". Consider introducing it spelled out here, as "integrity check value (ICV)". Later in the file, after these two `#define`, we can just use "ICV".